runtime/src/iree/hal/drivers/amdgpu/device/host_client.h

```diff
@@ -0,0 +1,115 @@
+enum iree_hal_amdgpu_device_host_client_call_e {
+  // Host will notify any registered listeners of the semaphore signal.
+  // The semaphore provided is a host handle to a generic HAL semaphore and may
+  // be of any device in the system - not just AMDGPU semaphores.
+  //
+  // Signature:
+  //   arg0: iree_hal_semaphore_t* semaphore
+  //   arg1: uint64_t payload
+  //   arg2: unused
+  //   arg3: unused
+  //   return_address: unused
+  //   completion_signal: unused
+  IREE_HAL_AMDGPU_DEVICE_HOST_CALL_POST_SIGNAL = 0u,
+
+  // Host will call iree_hal_resource_release on each non-NULL resource pointer.
+  // This is effectively a transfer operation indicating that the device will no
+  // longer be using the resources.
+  //
+  // It's strongly recommended that iree_hal_resource_set_t is used where
+  // appropriate so that the number of packets required to release a set of
+  // resources can be kept small. The 4 available here is just enough for the
+  // common case of submissions like execute that are a wait semaphore, the
+  // command buffer, the binding table resource set, and the signal semaphore.
+  //
+  // TODO(benvanik): evaluate a version that takes a ringbuffer of uint64_t
+  // pointers and make this a drain request instead. Then we can enqueue as many
+  // as we want and kick the host to drain as it is able.
+  //
+  // Signature:
+  //   arg0: iree_hal_resource_t* resource0
+  //   arg1: iree_hal_resource_t* resource1
+  //   arg2: iree_hal_resource_t* resource2
+  //   arg3: iree_hal_resource_t* resource3
+  //   return_address: unused
+  //   completion_signal: optional, signaled when the release has completed
+  IREE_HAL_AMDGPU_DEVICE_HOST_CALL_POST_RELEASE,
+};
+
+// Represents the host runtime thread that is managing host interrupts.
+// One or more schedulers may share a single host queue. An host calls that need
+// to identify the scheduler or scheduler-related resources must pass those as
+// arguments.
+typedef struct iree_hal_amdgpu_device_host_client_t {
+  // Host softqueue processing device requests. May be servicing requests from
+  // multiple device agents. Cached inline in device memory but references a
+  // ringbuffer in host memory.
+  iree_amd_cached_queue_t service_queue;
+  // Optional trace buffer used when tracing infrastructure is available.
+  iree_hal_amdgpu_device_trace_buffer_t* trace_buffer;
+} iree_hal_amdgpu_device_host_client_t;
+
+//===----------------------------------------------------------------------===//
+// Device-side Enqueuing
+//===----------------------------------------------------------------------===//
+
+#if defined(IREE_AMDGPU_TARGET_DEVICE)
+
+// Enqueues a unidirection host agent packet ("post").
+// Since this is device->host only operation this uses an acquire scope only
+// from the agent and releases to the entire system so the host agent can
+// observe changes. The completion signal is optional and may be
```

Comment:
I'm confused by the grammar here

```suggestion
// Since this is a device->host only operation this only uses an acquire scope
// from the agent and releases to the entire system so the host agent can
// observe changes. The completion signal is optional and may be
```

Reply:
me too :)